engine/src/test/java/io/zeebe/engine/state/WorkflowExecutionCleanZeebeStateTest.java

```diff
@@ -0,0 +1,618 @@
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .withElementType(BpmnElementType.PROCESS)
+        .await();
+
+    assertThatStateIsEmpty();
+  }
+
+  @Test
+  public void testWorkflowWithCallActivity() {
+    final var childWorkflow = Bpmn.createExecutableProcess("child").startEvent().endEvent().done();
+    final var parentWorkflow =
+        Bpmn.createExecutableProcess(PROCESS_ID)
+            .startEvent()
+            .callActivity("call", c -> c.zeebeProcessId("child"))
+            .endEvent()
+            .done();
+
+    engineRule
+        .deployment()
+        .withXmlResource("child.bpmn", childWorkflow)
+        .withXmlResource("parent.bpmn", parentWorkflow)
+        .deploy();
+
+    final var workflowInstanceKey =
+        engineRule.workflowInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    RecordingExporter.workflowInstanceRecords(WorkflowInstanceIntent.ELEMENT_COMPLETED)
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .withElementType(BpmnElementType.PROCESS)
+        .await();
+
+    assertThatStateIsEmpty();
+  }
+
+  @Test
+  public void testWorkflowCreatedWithResult() {
+    engineRule
+        .deployment()
+        .withXmlResource(Bpmn.createExecutableProcess(PROCESS_ID).startEvent().endEvent().done())
+        .deploy();
+
+    final var workflowInstanceKey =
+        engineRule
+            .workflowInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariable("x", 1)
+            .withResult()
+            .create();
+
+    RecordingExporter.workflowInstanceRecords(WorkflowInstanceIntent.ELEMENT_COMPLETED)
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .withElementType(BpmnElementType.PROCESS)
+        .await();
+
+    assertThatStateIsEmpty();
+  }
+
+  @Test
+  public void testWorkflowCanceled() {
+    engineRule
+        .deployment()
+        .withXmlResource(
+            Bpmn.createExecutableProcess(PROCESS_ID)
+                .startEvent()
+                .serviceTask("task", t -> t.zeebeJobType("test"))
+                .endEvent()
+                .done())
+        .deploy();
+
+    final var workflowInstanceKey =
+        engineRule.workflowInstance().ofBpmnProcessId(PROCESS_ID).withVariable("x", 1).create();
+
+    RecordingExporter.jobRecords(JobIntent.CREATED)
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .await();
+
+    engineRule.workflowInstance().withInstanceKey(workflowInstanceKey).cancel();
+
+    RecordingExporter.workflowInstanceRecords(WorkflowInstanceIntent.ELEMENT_TERMINATED)
+        .withWorkflowInstanceKey(workflowInstanceKey)
+        .withElementType(BpmnElementType.PROCESS)
+        .await();
+
+    assertThatStateIsEmpty();
+  }
+
+  private void assertThatStateIsEmpty() {
+    // sometimes the state takes few moments until is is empty
+    Awaitility.await()
+        .untilAsserted(
+            () -> {
+              assertThat(variablesState.isEmpty())
```

Comment:
Can we somehow just check whether the db is empty? I have the feeling this is a bit unstable if we miss to add a state here

Reply:
Yes. The tests depend on the fact that all states are checked and `isEmpty()` is correct. I don't see how to enforce it.

At least for now, we can not check if the whole RocksDB is empty because the deployments are not removed.

Reply:
hmpf..

What I would like to have is to check all column families and be able to exclude column families which should not be check like workflow column family. In that way we would find issues even if we add new column families, where resources are not cleaned up.

Reply:
Agree. Do you have an idea how we could do this?

Reply:
I have one. We can have something like this at the ZeebeDB impl.

```
  public boolean isEmpty(DbContext dbContext) {
    var isEmpty = false;
    for(var handle : columnFamilyMap.values()) {
      isEmpty &= isEmpty(handle, dbContext);
    }
    return isEmpty;
  }
```

But then you have the issue you meant before, with the column families where it is ok to contain values. I think we can modify the method in such a way that we ask for column families which contain values, like findNonEmptyColumnFamilies.  This method returns a list of column families which contain values. In the ZeebeState you can then check whether these families are the expected ones.

Reply:
Good idea. I introduced a new method in `ZeebeDb` to check if a column is empty.